File: services/backend/app/schemas/github.py
"""GitHub 账号、OAuth/device 与 import/push schema。"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class GitHubAccountOut(BaseModel):
    connected: bool
    login: str = ""
    name: str = ""
    avatar_url: str = ""
    scope: str = ""
    updated_at: datetime | None = None


class GitHubTokenConnectIn(BaseModel):
    token: str = Field(min_length=1, max_length=4096)


class GitHubOAuthStartOut(BaseModel):
    authorize_url: str


class GitHubDeviceStartIn(BaseModel):
    client_id: str | None = Field(default=None, max_length=128)
    scope: str = Field(default="repo", max_length=512)


class GitHubDeviceStartOut(BaseModel):
    device_code: str
    user_code: str
    verification_uri: str
    verification_uri_complete: str = ""
    expires_in: int
    interval: int


class GitHubDevicePollIn(BaseModel):
    client_id: str | None = Field(default=None, max_length=128)
    device_code: str = Field(min_length=1, max_length=512)


class GitHubDevicePollOut(BaseModel):
    status: str
    error: str = ""
    interval: int | None = None
    account: GitHubAccountOut | None = None


class GitHubImportIn(BaseModel):
    repo_url: str = Field(min_length=1, max_length=512)
    branch: str | None = Field(default=None, max_length=128)


class GitHubImportOut(BaseModel):
    project_id: str
    repo_url: str
    branch: str
    doc_count: int
    file_count: int
    byte_count: int


class GitHubPushIn(BaseModel):
    message: str | None = Field(default=None, max_length=512)


class GitHubPushOut(BaseModel):
    project_id: str
    repo_url: str
    branch: str
    commit_sha: str
    pushed: bool

File: services/frontend/src/services/backendApi/github.ts
/**
 * GitHub 账号、OAuth/device 与 import/push 相关 API。
 */

import { http } from './client'

export interface GitHubAccountStatus {
  connected: boolean
  login: string
  name: string
  avatar_url: string
  scope: string
  updated_at: string | null
}

export interface GitHubOAuthStart {
  authorize_url: string
}

export interface GitHubDeviceStart {
  device_code: string
  user_code: string
  verification_uri: string
  verification_uri_complete: string
  expires_in: number
  interval: number
}

export interface GitHubDevicePoll {
  status: 'pending' | 'slow_down' | 'connected' | 'failed' | string
  error: string
  interval: number | null
  account: GitHubAccountStatus | null
}

export interface GitHubImportResult {
  project_id: string
  repo_url: string
  branch: string
  doc_count: number
  file_count: number
  byte_count: number
}

export interface GitHubPushResult {
  project_id: string
  repo_url: string
  branch: string
  commit_sha: string
  pushed: boolean
}

export const githubApi = {
  account: () => http<GitHubAccountStatus>('/api/github/account', { scope: 'global' }),
  startOAuth: () =>
    http<GitHubOAuthStart>('/api/github/oauth/start', { method: 'POST', scope: 'global' }),
  startDevice: (clientId?: string) =>
    http<GitHubDeviceStart>('/api/github/device/start', {
      method: 'POST',
      body: JSON.stringify({ client_id: clientId || null, scope: 'repo' }),
      scope: 'global',
    }),
  pollDevice: (deviceCode: string, clientId?: string) =>
    http<GitHubDevicePoll>('/api/github/device/poll', {
      method: 'POST',
      body: JSON.stringify({ client_id: clientId || null, device_code: deviceCode }),
      scope: 'global',
    }),
  connectToken: (token: string) =>
    http<GitHubAccountStatus>('/api/github/token', {
      method: 'POST',
      body: JSON.stringify({ token }),
      scope: 'global',
    }),
  disconnect: () =>
    http<void>('/api/github/account', { method: 'DELETE', scope: 'global' }),
}
